crates/biome_js_analyze/src/lint/nursery/use_jsx_curly_brace_convention.rs

```diff
@@ -0,0 +1,339 @@
+    JsSyntaxToken, JsxAttributeInitializerClause, JsxExpressionAttributeValue, T,
+};
+use biome_rowan::{declare_node_union, AstNode, BatchMutationExt, TextRange};
+
+use crate::JsRuleAction;
+
+declare_rule! {
+    /// This rule allows you to enforce curly braces or disallow unnecessary curly braces in JSX props and/or children.
+    ///
+    /// For situations where JSX expressions are unnecessary, please refer to [the React doc](https://facebook.github.io/react/docs/jsx-in-depth.html) and [this page about JSX gotchas](https://github.com/facebook/react/blob/v15.4.0-rc.3/docs/docs/02.3-jsx-gotchas.md#html-entities).
+    ///
+    /// By default, this rule will check for and warn about unnecessary curly braces in both JSX props and children.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```jsx,expect_diagnostic
+    /// <Foo>{'Hello world'}</Foo>;
+    ///
+    /// <Foo foo={'bar'} />;
+    ///
+    /// <Foo foo=<Bar /> />;
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// <Foo>Hello world</Foo>;
+    /// <Foo foo="bar" />;
+    /// <Foo foo={5} />;
+    /// <Foo foo={<Bar />} />;
+    /// ```
+    ///
+    pub UseJsxCurlyBraceConvention {
+        version: "next",
+        name: "useJsxCurlyBraceConvention",
+        language: "js",
+        recommended: false,
+        sources: &[RuleSource::EslintReact("jsx-curly-brace-presence")],
+        source_kind: RuleSourceKind::Inspired,
+        fix_kind: FixKind::Safe,
+    }
+}
+
+declare_node_union! {
+    pub UseJsxCurlyBraceConventionQuery = JsxAttributeInitializerClause | AnyJsxChild
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CurlyBraceResolution {
+    /// The user should add curly braces around the expression.
+    AddBraces,
+    /// The user should remove the curly braces around the expression.
+    RemoveBraces,
+}
+
+impl Rule for UseJsxCurlyBraceConvention {
+    type Query = Ast<UseJsxCurlyBraceConventionQuery>;
+    type State = CurlyBraceResolution;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let query = ctx.query();
+        let has_curly_braces = has_curly_braces(query);
+        match query {
+            UseJsxCurlyBraceConventionQuery::JsxAttributeInitializerClause(attr) => {
+                handle_attr_init_clause(attr, has_curly_braces)
+            }
+            UseJsxCurlyBraceConventionQuery::AnyJsxChild(child) => {
+                handle_jsx_child(child, has_curly_braces)
+            }
+        }
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+        let source_range = match &node {
+            UseJsxCurlyBraceConventionQuery::JsxAttributeInitializerClause(node) => {
+                node.value().map(|value| value.range())
+            }
+            _ => Ok(node.range()),
+        }
+        .unwrap_or(node.range());
```

Comment:
You should `impl UseJsxCurlyBraceConventionQuery` and add a `source_range` method instead and document it. That's what we do throughout the code base. From this method, you should return a `Option<TextRange>`, write something like this:

```rs

fn source_range(&self) -> Option<TextRange> {
	match self {
		UseJsxCurlyBraceConventionQuery::JsxAttributeInitializerClause(node) => {
			let value = node.value()?;
			Some(value.range())
		},
		UseJsxCurlyBraceConventionQuery::AnyJsxChild(node) => Some(node.range())
	}
}
```

I suggest avoiding `_ => {}` as much as possible because if in the future you add a new variant to an enum, the compiler won't warn you, and you might forget some implementations